vmm: support stratovirt and qemu sandbox recovery

Vanient opened the pull request:
Recover stratovirt or qemu sandbox information when vmm-sandboxer restarting to make sure sandbox started before could be controlled.

flyflypeng commented:
I have refactored the vmm/sandbox/src/main.rs#main function to remove the repeated code in the #36.
This PR may rebase after the #36 PR merged.

Vanient commented:
> I have refactored the vmm/sandbox/src/main.rs#main function to remove the repeated code in the #36. This PR may rebase after the #36 PR merged.

rebased

flyflypeng commented on vmm/sandbox/src/stratovirt/mod.rs:
```diff
@@ -505,3 +499,18 @@ impl StratoVirtVM {
+
+#[async_trait::async_trait]
+impl Recoverable for StratoVirtVM {
+    async fn recover(&mut self) -> Result<()> {
```
There are repeated code in the `recover` function with different hypervisor type, maybe we can extract the same code block into trait function implementation.

Vanient replied:
fixed

Burning1020 reviewed (approved):
LGTM

abel-von commented on vmm/sandbox/src/stratovirt/mod.rs:
```diff
@@ -69,29 +70,22 @@ pub(crate) const STRATOVIRT_START_TIMEOUT_IN_SEC: u64 = 10;
     #[serde(skip)]
     devices: Vec<Box<dyn StratoVirtDevice + Sync + Send>>,
     #[serde(skip)]
     hot_attached_devices: Vec<Box<dyn StratoVirtHotAttachable + Sync + Send>>,
```
the hot_attached_devices for qemu should be recovered when restart, otherwise the devices of the container will be residual after the container is restarted.

abel-von replied:
But I found that the hot_detach is not implemented for stratovirt, how to detach the devices from vm after a container stopped?

Vanient replied:
Yes, stratovirt vm hot_detach is not implemented yet, which may cause residual problems as you say, especially when using dm as storage driver. But as lack of device recovery has no influence on basic sandbox and container management on overlay scene, and it is a little complicated, is it ok if I open a new issue to track device recovery, not implement in this commit?

abel-von commented on vmm/sandbox/src/stratovirt/mod.rs:
```diff
@@ -69,29 +70,22 @@ pub(crate) const STRATOVIRT_START_TIMEOUT_IN_SEC: u64 = 10;
-    #[serde(skip)]
     virtiofs_daemon: Option<VirtiofsDaemon>,
     #[serde(skip)]
     pcie_root_bus: PcieRootBus,
```
We need to check carefully of these fields, if it is skipped from serialization, it should also be rebuild in some other way, otherwise operations to the VM may failed. the this pcie_root_bus for example, if the value is not recovered, how can we hot attach a new device to the vm, if a new container is created in the sandbox?

Vanient replied:
fixed

abel-von commented on vmm/sandbox/src/vm.rs:
```diff
@@ -72,6 +72,26 @@ pub trait VM: Serialize + Sync + Send {
+            }
+        }
+    };
+}
```
Based on the comments above, I don't think the recovering process must be very different from each other, if it is so, maybe extracting this macro is not necessary. 

If the recovering is so common, that the recover codes is duplicated, maybe there is another way to do it.
```
impl<T: VM> Recoverable for T {
    async fn recover(&mut self) -> Result<()>
}

```

and add `create_client` and `pid` to the trait of VM, so that the compiler could easily checkout the constraint that we can only recover the object with `create_client` and `pid` method.

abel-von commented on vmm/sandbox/src/qemu/mod.rs:
```diff
@@ -70,21 +71,15 @@ pub(crate) const QEMU_START_TIMEOUT_IN_SEC: u64 = 10;
     #[serde(skip)]
     devices: Vec<Box<dyn QemuDevice + Sync + Send>>,
     #[serde(skip)]
     hot_attached_devices: Vec<Box<dyn QemuHotAttachable + Sync + Send>>,
```
the hot_attached_devices should be recovered when restart, otherwise the devices of the container will be residual after the container is restarted.

abel-von commented on vmm/sandbox/src/qemu/mod.rs:
```diff
@@ -70,21 +71,15 @@ pub(crate) const QEMU_START_TIMEOUT_IN_SEC: u64 = 10;
-    #[serde(skip)]
     netns: String,
     #[serde(skip)]
     block_driver: BlockDriver,
```
maybe we also have to recover this field so that we can decide which block driver to use for block device hot attach?

Vanient replied:
fixed
